evap/rewards/tools.py

```diff
@@ -71,14 +84,21 @@ def grant_reward_points(user, semester):
     required_courses = Course.objects.filter(participants=user, semester=semester, is_required_for_reward=True)
     if not required_courses.exists():
         return False
-    # does the user not participate in any more required courses in this semester?
-    if required_courses.exclude(voters=user).exists():
-        return False
-    # did the user not already get reward points for this semester?
-    if RewardPointGranting.objects.filter(user_profile=user, semester=semester).exists():
+
+    voted_courses_count = required_courses.filter(voters=user).count()
+    # full evaluation progress from 0.0 to 1.0
+    progress = float(voted_courses_count) / float(required_courses.count())
+
+    # How many points have been granted to this user this semester?
+    granted_points = RewardPointGranting.objects.filter(user_profile=user, semester=semester).aggregate(Sum('value'))['value__sum'] or 0
+    points_missing = target_points(progress) - granted_points
```

Comment:
to be honest, i would inline the whole thing and simply do
`
target_points = max(points for threshold, points in settings.REWARD_POINTS if threshold <= progress, default=0)
`
target points are the maximum of all point values whose threshold was passed, or zero if none was passed. in my opinion that is clear enough, and adding a 10 line method for this might add more overhead when reading than it's helping.

Reply:
...of course this method cannot be tested then anymore, but i think testing the whole view is just fine.